ICSharpCode.Decompiler/IL/ControlFlow/ConditionDetection.cs

```diff
@@ -47,264 +59,546 @@ public class ConditionDetection : IBlockTransform
+		///   ...; br nextBlock;
+		/// ->
+		///   ...; { nextBlock... }
+		/// 
+		/// Only inlines branches that are strictly dominated by this block (incoming edge count == 1)
+		/// </summary>
+		private bool InlineExitBranch(Block block)
+		{
+			var exitInst = GetExit(block);
+			if (!CanInline(exitInst))
+				return false;
+
+			context.Step("Inline target block of unconditional branch", exitInst);
+			// The targetBlock was already processed, and is ready to embed
+			var targetBlock = ((Branch)exitInst).TargetBlock;
+			block.Instructions.RemoveAt(block.Instructions.Count - 1);
+			block.Instructions.AddRange(targetBlock.Instructions);
+			targetBlock.Remove();
+
+			return true;
+		}
+
+		/// <summary>
+		/// Gets whether <c>potentialBranchInstruction</c> is a branch to a block that is dominated by <c>cfgNode</c>.
+		/// If this function returns true, we replace the branch instruction with the block itself.
+		/// </summary>
+		private bool CanInline(ILInstruction exitInst)
+		{
+			if (exitInst is Branch branch
+			    && branch.TargetBlock.Parent == currentContainer
+			    && branch.TargetBlock.IncomingEdgeCount == 1
+			    && branch.TargetBlock.FinalInstruction is Nop) {
+				// if the incoming edge count is 1, then this must be the sole branch, and dominance is already ensured
+				Debug.Assert(cfgNode.Dominates(context.ControlFlowGraph.GetNode(branch.TargetBlock)));
+				return true;
 			}
+
+			return false;
 		}
 
-		private void HandleIfInstruction(ControlFlowNode cfgNode, Block block, IfInstruction ifInst, ref ILInstruction exitInst)
+		/// <summary>
+		/// Looks for common exits in the inlined then and else branches of an if instruction
+		/// and performs inversions and simplifications to merge them provided they don't 
+		/// isolate a higher priority block exit
+		/// </summary>
+		private void MergeCommonBranches(Block block, IfInstruction ifInst)
 		{
-			if (ShouldSwapIfTargets(ifInst.TrueInst, exitInst)) {
-				// "if (c) goto lateBlock; goto earlierBlock;"
-				// -> "if (!c)" goto earlierBlock; goto lateBlock;
-				// This reordering should make the if structure correspond more closely to the original C# source code
-				context.Step("Negate if", ifInst);
-				block.Instructions[block.Instructions.Count - 1] = ifInst.TrueInst;
-				ifInst.TrueInst = exitInst;
-				exitInst = block.Instructions.Last();
-				ifInst.Condition = Comp.LogicNot(ifInst.Condition);
+			var thenExits = new List<ILInstruction>();
+			AddExits(ifInst.TrueInst, 0, thenExits);
+			if (thenExits.Count == 0)
+				return;
+
+			// if there are any exits from the then branch, then the else is redundant and shouldn't exist
+			Debug.Assert(IsEmpty(ifInst.FalseInst));
+			var elseExits = new List<ILInstruction>();
+			int falseInstIndex = block.Instructions.IndexOf(ifInst) + 1;
+			AddExits(block, falseInstIndex, elseExits);
+			
+			var commonExits = elseExits.Where(e1 => thenExits.Any(e2 => DetectExitPoints.CompatibleExitInstruction(e1, e2)));
```

Comment:
correct, clearer code first, optimisations if obvious or profiling hotspot, generally not that many exits (especially since it's only looking for exits which could be moved to the block root via inversions.